Discourage fee sniping with nLockTime

danielabrozzoni opened the pull request:
### Description
By default bdk sets the transaction's nLockTime to current_height
to prevent fee sniping.
current_height can be provided by the user through TxParams; if the user
didn't provide it, we use the last sync height, or 0 if we never synced.

Fixes https://github.com/bitcoindevkit/bdk/issues/533

### Notes to the reviewers:

If you want to know more about fee sniping: https://bitcoinops.org/en/topics/fee-sniping/

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature
* [x] I've updated `CHANGELOG.md`

danielabrozzoni committed:
Discourage fee sniping with nLockTime

By default bdk sets the transaction's nLockTime to current_height
to discourage fee sniping.
current_height can be provided by the user through TxParams; if the user
didn't provide it, we use the last sync height, or 0 if we never synced.

Fixes #533

afilini commented on src/wallet/tx_builder.rs:
```diff
@@ -542,6 +543,15 @@ impl<'a, D: BatchDatabase, Cs: CoinSelectionAlgorithm<D>, Ctx: TxBuilderContext>
+    ///
+    /// This will be used to set the nLockTime for preventing fee sniping. If the current height is
+    /// not provided, the last sync height will be used instead.
+    pub fn set_current_height(&mut self, height: BlockTime) -> &mut Self {
```
I would use `u32` so that we don't have to create a `BlockTime` struct with an unused timestamp

rajarshimaitra commented on src/wallet/mod.rs:
```diff
@@ -1890,6 +1906,76 @@ pub(crate) mod test {
+        let addr = wallet.get_address(New).unwrap();
+        let mut builder = wallet.build_tx();
+        builder.add_recipient(addr.script_pubkey(), 25_000);
+        let sync_height = SyncTime {
```
Nit: Could be called `sync_time` just to make naming consistency..

Ignore if you don't touch it again..

rajarshimaitra commented on src/wallet/mod.rs:
```diff
@@ -1890,6 +1906,76 @@ pub(crate) mod test {
+    }
+
+    #[test]
+    fn test_create_tx_fee_sniping_locktime_sync_height() {
```
Maybe I am confused somewhere, whats the difference between this test and the test before this? It seems both are checking the same thing with just different value?

danielabrozzoni replied:
Right, I have no idea why I did this... I probably wanted to test the "never synced" case, and then I got confused 🤦🏻‍♀️ Sorry!

rajarshimaitra commented on src/wallet/mod.rs:
```diff
@@ -1890,6 +1906,76 @@ pub(crate) mod test {
+        let addr = wallet.get_address(New).unwrap();
+        let mut builder = wallet.build_tx();
+        builder.add_recipient(addr.script_pubkey(), 25_000);
+        let sync_height = SyncTime {
```
Same naming nit..

rajarshimaitra commented on src/wallet/mod.rs:
```diff
@@ -613,9 +613,25 @@ where
         let lock_time = match params.locktime {
-            // No nLockTime, default to 0
-            None => requirements.timelock.unwrap_or(0),
+            // No nLockTime specified, we try to prevent fee sniping, if possible
```
Nit: 
```suggestion
            // When no nLockTime is specified, we try to prevent fee sniping, if possible
```

rajarshimaitra commented on src/wallet/mod.rs:
```diff
@@ -613,9 +613,25 @@ where
+            None => {
+                // Fee sniping can be partially prevented by setting the timelock
+                // to current_height. If we don't know the current_height,
+                // we have to default to 0.
```
Nit: Not like we "have" to, but we just choose it.. 😛

rajarshimaitra reviewed (commented):
tACK f8da54e9eca8bad65444f0278a077dc063acef30

Just for the sake of completeness, maybe one more test showing, when the wallet isn't synced locktime defaults to 0?

And few other non blocking nits..

danielabrozzoni commented:
Thanks for the review! I updated the code removing the duplicated test, and adding some checks to the already existing tests (for example, testing that the current_height won't override the user specified nlocktime).
Turns out that the "never synced" case was already tested, in `test_create_tx_default_locktime` - but I added a comment to explicit this.

rajarshimaitra commented on src/wallet/tx_builder.rs:
```diff
@@ -147,6 +147,7 @@ pub(crate) struct TxParams {
     pub(crate) add_global_xpubs: bool,
     pub(crate) include_output_redeem_witness_script: bool,
     pub(crate) bumping_fee: Option<PreviousFee>,
+    pub(crate) current_height: Option<u32>,
```
I feel it would be really helpful for lib users if these fields can come with their comments explaining what are they for.. 

Doesn't have to be included in this PR. But something for future..

danielabrozzoni replied:
When it's going to be exposed publicly, sure :) For now just the devs are seeing it, and I'd say it's easy enough to understand the various params, since the functions modifying them are commented anyways...

rajarshimaitra replied:
It has a TODO to be exposed publicly, and as internal dev usage also I feel its better to document them to disclose their intents.. 

But yes not blocking on this PR. Can be done later..

rajarshimaitra reviewed (approved):
tACK dc5471fc405f21cea43462d3cbd2071ad98ac92e

Thanks for the update.. Just one more comment for future PR maybe.. 

This one seems good to go..

danielabrozzoni commented:
Rebased

afilini commented on CHANGELOG.md:
```diff
@@ -18,6 +18,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   - Creation of Taproot PSBTs (BIP-371)
   - Signing Taproot PSBTs (key spend and script spend)
   - Support for `tr()` descriptors in the `descriptor!()` macro
+- Transaction nlocktime defaults to `current_height`, which can be specified through `TxParams` - if not specified, the last sync height is used
```
This is a bit confusing, what are you trying to say exactly? It sounds like the default is `current_height`, but then you say that it can be specified manually (so it's not a default ?), and then again if it's not specified something else happens.

afilini commented on src/wallet/mod.rs:
```diff
@@ -619,9 +619,25 @@ where
+            .map(|sync_time| sync_time.block_time.height);
+
+        // If they didn't tell us the current height, we assume it's the latest sync height.
+        let current_height = params.current_height.or(last_sync_height);
```
It's not the end of the world, but maybe you could move the database lookup inside an `or_else` closure, so that it happens only if it's actually required. Otherwise you may fetch the last sync height and then throw it away because there was a custom `curent_height` set.

danielabrozzoni replied:
While this is technically possible, it's not really ideal:

- Either I completely discard whatever error get_sync_time throws to me:
```
       let current_height = params.current_height.or_else(|| {
           self.database()
               .get_sync_time()
               .ok()?
               .map(|sync_time| sync_time.block_time.height)
       });

```

- Or I turn this simple expression into a functional mess:
```
        let current_height = params.current_height.map_or_else(|| -> Result<_, Error> {
            Ok(self.database()
                .get_sync_time()?
                .map(|sync_time| sync_time.block_time.height))
        }, |x| Ok(Some(x)))?;
```

afilini commented on src/wallet/mod.rs:
```diff
@@ -619,9 +619,25 @@ where
+                // we default to 0.
+                let fee_sniping_height = current_height.unwrap_or(0);
+                // We choose the bigger between the required height and the fee sniping
+                // height
```
```suggestion
                // We choose the biggest between the required nLockTime and the fee sniping
                // height
```

afilini commented on src/wallet/tx_builder.rs:
```diff
@@ -543,6 +544,15 @@ impl<'a, D: BatchDatabase, Cs: CoinSelectionAlgorithm<D>, Ctx: TxBuilderContext>
+    /// Set the current blockchain height.
+    ///
+    /// This will be used to set the nLockTime for preventing fee sniping. If the current height is
+    /// not provided, the last sync height will be used instead.
```
Maybe you should document that `TxParams::locktime()` always takes precedence over this?

vladimirfomene commented on src/wallet/mod.rs:
```diff
@@ -619,9 +619,25 @@ where
         let lock_time = match params.locktime {
-            // No nLockTime, default to 0
-            None => requirements.timelock.unwrap_or(0),
+            // When no nLockTime is specified, we try to prevent fee sniping, if possible
+            None => {
+                // Fee sniping can be partially prevented by setting the timelock
+                // to current_height. If we don't know the current_height,
+                // we default to 0.
+                let fee_sniping_height = current_height.unwrap_or(0);
```
Tested ACK. Pulled the code locally, tested on branch, merged on master and tested. @danielabrozzoni I know this PR passed the design phase but I was wondering why on fee sniping prevention you are setting the timelock to the `current_height ` and not `current_height + 1` as in Bitcoin Core ([check last paragraph of this section](https://github.com/bitcoinbook/bitcoinbook/blob/develop/ch07.asciidoc#timelock-defense-against-fee-sniping)).

danielabrozzoni replied:
That's an excellent question! I have to admit that when I started working on this, I was pretty sure Core would set it to `current_height + 1`. After implementing it tho, I noticed that if I set it to `current_height + 1`, then nodes wouldn't relay my transaction (as the nLockTime wasn't expired yet)!
After digging a bit into Bitcoin Core, I found out that they, as well, set the nLockTime to the best height:

https://github.com/bitcoin/bitcoin/blob/45d8b1e94a8e3e68f210d496330f5e16219d2a8d/src/wallet/spend.cpp#L633

But then, they improve privacy by setting the nLockTime even further back in the past sometimes:
https://github.com/bitcoin/bitcoin/blob/45d8b1e94a8e3e68f210d496330f5e16219d2a8d/src/wallet/spend.cpp#L639-L641

Seems like Mastering Bitcoin is wrong... There's already an issue open https://github.com/bitcoinbook/bitcoinbook/issues/997, I hope they'll eventually fix :)

vladimirfomene replied:
Thank you for sharing your learnings with me!

vladimirfomene reviewed (commented):
Tested ACK. Just had a question

kafaichoi commented on src/wallet/mod.rs:
```diff
@@ -619,9 +619,25 @@ where
+            .map(|sync_time| sync_time.block_time.height);
+
+        // If they didn't tell us the current height, we assume it's the latest sync height.
+        let current_height = params.current_height.or(last_sync_height);
```
should we use `or_else` and inline the function calls here as this will be lazy evaluated. So no db lookup is performed if current_height is provided

danielabrozzoni replied:
Yeah, it's not that easy, see https://github.com/bitcoindevkit/bdk/pull/611#discussion_r892192599
Let me know if you have a better way of implementing it 😄

kafaichoi replied:
ahh sorry I didn't see there is an existing comment. Maybe using match might be slightly more clear in this case?

```
        let current_height = match params.current_height {
            None => self.database().get_sync_time()?.map(|sync_time| sync_time.block_time.height),
            h => h
        };

```

danielabrozzoni replied:
Oh that's cool, I'll update the code, thanks!

danielabrozzoni replied:
Updated, thanks!

rajarshimaitra reviewed (approved):
ReACK 07c77f95ee5316c76daade8a8c26934ec2262a62

afilini commented:
This needs rebasing and then it's ready to go

danielabrozzoni commented:
Rebased :)

afilini reviewed (approved):
ACK 97bc9dc7170c336e97cd756b1f07ac3c23a39626
